Shark Cage/createACLinNewDesktop/createACLinNewDesktop/util.cpp

```diff
@@ -4,121 +4,121 @@
 #include <lmerr.h>
 #include <stdexcept>
 #include <Sddl.h>
-//#include <Wdm.h>
-//#include <Ntddk.h>
-//#include <Ntifs.h>
-#include <Ntdef.h>
 #include <Wtsapi32.h>
-
-//#include <ntsecapi.h>
-
+#include <Winternl.h>
+#include <ntstatus.h>
 #include <iostream>
 #include <stdio.h>
 #pragma comment(lib, "netapi32.lib")
-//#pragma comment(lib, "Wtsapi32.lib")
 
 class tokenStructures {
 public:
-	ACCESS_MASK			AccessMask; //OK
-	POBJECT_ATTRIBUTES	ObjectAttributes; //OK
-	TOKEN_TYPE			TokenType; //OK
-	PLUID				AuthenticationId; //OK
-	PLARGE_INTEGER		ExpirationTime; //OK
-	PTOKEN_USER         TokenUser; //OK
-	PTOKEN_GROUPS       TokenGroups; //OK
-	PTOKEN_PRIVILEGES   TokenPrivileges; //OK
-	PTOKEN_OWNER        TokenOwner; //OK
-	PTOKEN_PRIMARY_GROUP TokenPrimaryGroup; //OK
-	PTOKEN_DEFAULT_DACL TokenDefaultDacl; //OK
-	PTOKEN_SOURCE       TokenSource; //OK
+	ACCESS_MASK			AccessMask;
+	POBJECT_ATTRIBUTES	ObjectAttributes;
+	TOKEN_TYPE			TokenType;
+	PLUID				AuthenticationId;
+	PLARGE_INTEGER		ExpirationTime;
+	PTOKEN_USER         TokenUser;
+	PTOKEN_GROUPS       TokenGroups;
+	PTOKEN_PRIVILEGES   TokenPrivileges;
+	PTOKEN_OWNER        TokenOwner;
+	PTOKEN_PRIMARY_GROUP TokenPrimaryGroup;
+	PTOKEN_DEFAULT_DACL TokenDefaultDacl;
+	PTOKEN_SOURCE       TokenSource;
 };
 
-bool emergencyExit(LPTSTR pUser_name); 
 void enumerateSidsAndHashes(PTOKEN_ACCESS_INFORMATION pToken);
 ULONG getCurrentSessionID();
 void deconstructToken(tokenStructures &tokenDeconstructed, HANDLE &userToken);
+bool changeTokenCreationPrivilege(bool privilegeStatus);
+bool addGroupToTokenGroups(PSID sid, tokenStructures &tokenDeconstructed);
 
 
 
-namespace util {
-
-	LPWSTR  group_name = L"Dummy_group"; //TODO: this should be constant, fix the conversions
-
-
-	//this is all obsolete now  - keeping this only for grading purposes
-	//reason: there is no way to arbitrarilly add a group to the token
-	//we are going to create a new token from scratch using NtCreateToken
-	bool getModifiedToken(PSID sid, HANDLE &token) {
-
+namespace tokenLib {
 
+	bool createLocalGroup(LPWSTR groupName, PSID &sid){
+		SID_NAME_USE accountType;
 		LOCALGROUP_INFO_0 localgroup_info;
-		LPTSTR   pUser_name = NULL;
-		HANDLE current_process_handle;
-		HANDLE user_token_h;
-		DWORD bufferSize = 0;
-
+		localgroup_info.lgrpi0_name = groupName;
 
-		//create a dummy group
-		localgroup_info.lgrpi0_name = group_name;
-		if (NetLocalGroupAdd(NULL, 0, (LPBYTE)&localgroup_info, NULL) != NERR_Success)
-			return emergencyExit(pUser_name);
+		NetLocalGroupDel(NULL, groupName);

```

Comment:
isn't it kind of dangerous to just delete the group by its name? what happens if a group with this name already exists on the machine? (unlikely, but who knows...)

Reply:
Well... the idea is that the group must be unique (i believe group names are unique as well) and randomized. If the groupname we would like to use already exists, our security model is broken.  The other thing is that the way you create group is by providing a name. If you need an sid, you need to perform a manual lookup *by name*. The other way to handle this situation is to return error code saying something along the lines "this groupname is already in use". Would you consider this better?

Reply:
You are of course correct in stating that we should never use a group name that already exists, but I still think it would be wise to check an fail appropriately here in case someone on the caller side forgets / doesn't know what happens here (who knows how the token lib gets used in the future)

Reply:
Fixed, function now returns false and error message if a name conflict is encountered